mypyc/irbuild/for_helpers.py

```diff
@@ -436,12 +439,32 @@ def make_for_loop_generator(
 
     rtyp = builder.node_type(expr)
     if is_sequence_rprimitive(rtyp):
-        # Special case "for x in <list>".
-        expr_reg = builder.accept(expr)
+        # Special case "for x in <list/tuple/str/bytes>".
         target_type = builder.get_sequence_type(expr)
-
         for_list = ForSequence(builder, index, body_block, loop_exit, line, nested)
-        for_list.init(expr_reg, target_type, reverse=False)
+
+        if isinstance(expr, IndexExpr) and isinstance(expr.index, SliceExpr):
+            # TODO: maybe we must not apply this optimization to list type specifically
+            # because the need to check length changes at each iteration?
```

Comment:
This can probably still be valid for ListComp and ListExpr 

```python
def do_thing():
    return [x + 4 for x in some_non_reversible_generator][1:-2:-1]
```

not sure how useful that is though